internal/nfs/nodeserver/nodeserver.go

```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeserver
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	csicommon "github.com/ceph/ceph-csi/internal/csi-common"
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	mount "k8s.io/mount-utils"
+	netutil "k8s.io/utils/net"
+)
+
+const (
+	defaultMountPermission = uint64(0o777)
+	// Address of the NFS server.
+	paramServer           = "server"
+	paramShare            = "share"
+	paramClusterID        = "clusterID"
+	mountOptionsField     = "mountOptions"
+	mountPermissionsField = "mountPermissions"
+)
+
+// NodeServer struct of ceph CSI driver with supported methods of CSI
+// node server spec.
+type NodeServer struct {
+	*csicommon.DefaultNodeServer
+	Mounter mount.Interface
+}
+
+// NewNodeServer initialize a node server for ceph CSI driver.
+func NewNodeServer(
+	d *csicommon.CSIDriver,
+	t string,
+	topology map[string]string,
+) *NodeServer {
+	defaultNodeServer := csicommon.NewDefaultNodeServer(d, t, topology)
+
+	return &NodeServer{
+		DefaultNodeServer: defaultNodeServer,
+		Mounter:           defaultNodeServer.Mounter,
+	}
+}
+
+// NodePublishVolume mount the volume.
+func (ns *NodeServer) NodePublishVolume(
+	ctx context.Context,
+	req *csi.NodePublishVolumeRequest,
+) (*csi.NodePublishVolumeResponse, error) {
+	err := validateNodePublishVoluemRequest(req)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	volumeID := req.GetVolumeId()
+	volCap := req.GetVolumeCapability()
+	targetPath := req.GetTargetPath()
+
+	// Considering kubelet make sure the stage and publish operations
+	// are serialized, we dont need any extra locking in nodePublish.
+
+	mountOptions := volCap.GetMount().GetMountFlags()
+	if req.GetReadonly() {
+		mountOptions = append(mountOptions, "ro")
+	}
+
+	var server, baseDir, clusterID string
+
+	mountPermissions := defaultMountPermission
+	for k, v := range req.GetVolumeContext() {
+		switch k {
+		case paramServer:
+			server = v
+		case paramShare:
+			baseDir = v
+		case paramClusterID:
+			clusterID = v
+		case mountOptionsField:
+			if v != "" {
+				mountOptions = append(mountOptions, v)
+			}
+		case mountPermissionsField:
```

Comment:
do we have this field set by storageclass or the controller?

Reply:
it should be passed through storageclass params, I'll add docs for the same.

Reply:
are docs missing in the csi-driver-nfs that has this option? Please add a PR for it there as well then.